homeassistant/components/vodafone_station/api.py

```diff
@@ -0,0 +1,396 @@
+    def wifi(self) -> str | None:
+        """Return device WIFi connection."""
+        return self._wifi
+
+
+class VodafoneStationApi(DataUpdateCoordinator):
+    """Queries router running Vodafone Station firmware."""
+
+    def __init__(
+        self, host: str, ssl: bool, username: str, password: str, hass: HomeAssistant
+    ) -> None:
+        """Initialize the scanner."""
+        super().__init__(
+            hass=hass,
+            logger=_LOGGER,
+            name=f"{DOMAIN}-{host}-coordinator",
+            update_interval=timedelta(seconds=30),
+        )
+
+        self.host = host
+        self.protocol = "https" if ssl else "http"
+        self.username = username
+        self.password = password
+        self.hass = hass
+        self.base_url = f"{self.protocol}://{self.host}"
+        self.headers = {
+            "User-Agent": "Mozilla/5.0 (X11; Fedora; Linux x86_64; rv:78.0) Gecko/20100101 Firefox/78.0",
+            "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8",
+            "Accept-Language": "en-GB,en;q=0.5",
+            "Origin": self.base_url,
+            "Referer": f"{self.base_url}/login.html",
+            "DNT": "1",
+        }
+        jar = aiohttp.CookieJar(unsafe=True)
+        self.session = aiohttp.ClientSession(cookie_jar=jar)
+        self.csrf_token: str = ""
+        self.encryption_key: str = ""
+        self._unique_id: str | None = None
+        self._devices: dict[str, VodafoneStationDeviceInfo] = {}
+        self._data: dict[str, str] = {}
+        self._on_close: list[Callable] = []
+
+    async def _get_csrf_token(self) -> None:
+        """Load login page to get csrf token."""
+
+        url = f"{self.base_url}/login.html"
+        reply = await self.session.get(
+            url,
+            headers=self.headers,
+            timeout=10,
+            verify_ssl=False,
+            allow_redirects=False,
+        )
+        reply_text = await reply.text()
+        tokens = re.search("(?<=csrf_token = ')[^']+", reply_text)
+        if not tokens:
+            return None
+        self.csrf_token = tokens.group(0)
+        _LOGGER.debug("csrf_token: <%s>", self.csrf_token)
+
+    async def _get_user_lang(self) -> None:
+        """Load user_lang page to get."""
+
+        timestamp = datetime.now().strftime("%s")
+        url = f"{self.base_url}/data/user_lang.json?_={timestamp}&csrf_token={self.csrf_token}"
+        reply = await self.session.get(
+            url,
+            headers=self.headers,
+            timeout=10,
+            verify_ssl=False,
+            allow_redirects=False,
+        )
+
+        j = await reply.json(content_type="text/html")
+        user_obj = {}
+        for item in j:
+            key = list(item.keys())[0]
+            val = list(item.values())[0]
+            user_obj[key] = val
+
+        self.encryption_key = user_obj["encryption_key"]
+        _LOGGER.debug("encryption_key: <%s>", self.encryption_key)
+
```

Comment:
Most of this belongs in an external library

Reply:
moved to aiovodafone